lib/charms/grafana_agent_k8s/v0/grafana_agent.py

```diff
@@ -155,37 +192,88 @@ def _on_log_proxy_relation_departed(self, event):
         self._update_agents_list(event)
 
         if len(self._current_config["clients"]) == 0:
-            self._container.stop(self._container_name)
+            self._container.stop("promtail")
         else:
-            self._container.restart(self._container_name)
+            self._container.restart("promtail")
 
     def _on_upgrade_charm(self, event):
         # TODO: Implement it ;-)
         pass
 
-    def _on_promtail_pebble_ready(self, _):
-        """Event handler for the pebble ready event.
+    def _get_container(self, container_name):
+        if container_name is not None:
+            return self._charm.unit.get_container(container_name)
 
-        Args:
-            event: The event object of the pebble ready event
-        """
+        containers = dict(self._charm.model.unit.containers)
+
+        if len(containers) == 1:
+            return self._charm.unit.get_container([*containers].pop())
+
+        # FIXME: Use custom exception
+        raise Exception("Container cannot be obtained")
 
+    def _build_pebble_layer(self):
+        """Event handler for the pebble ready event."""
         pebble_layer = {
             "summary": "promtail layer",
             "description": "pebble config layer for promtail",
             "services": {
                 "promtail": {
                     "override": "replace",
                     "summary": "promtail",
-                    "command": f"/usr/bin/promtail {self._cli_args}",
-                    "startup": "disable",
+                    "command": "{} {}".format(WORKLOAD_BINARY_PATH, self._cli_args),
+                    "startup": "enabled",
                 }
             },
         }
+        self._container.add_layer(self._container_name, pebble_layer, combine=True)
+
+    def _obtain_promtail(self, event) -> None:
+        if self._is_promtail_binary_in_workload():
+            return
+
+        self._download_promtail(event)
+
+        if not self._check_sha256sum():
+            msg = "Promtail bnary sha256sum mismatch"
+            logger.error(msg)
+            raise PromtailDigestError(msg)
+
+        self._unzip_binary()
+        self._upload_binary()
```

Comment:
The number of these little 1-2 line helpers makes this method kinda awkward to read imo -- consider the following:

```python
from shutil import copyfileobj

def _download_promtail(self, event) -> None:
    # Fetch the binary
    response = urlopen(PROMTAIL_BINARY_ZIP_URL)
    with response as r, open(BINARY_ZIP_PATH, "wb") as f:
        copyfileobj(r, f)

    # Validate that it's legit
    with open(BINARY_ZIP_PATH, "rb") as f:
        result = sha256(f.read()).hexdigest()

    if BINARY_SHA256SUM != result:
        logger.error(
            "promtail binary mismatch, expected:'{}' but got '{}', 
            BINARY_SHA256SUM, result)
        raise PromtailDigestError("Digest mismatch for promtail binary")

    # Unzip the binary
    with ZipFile(BINARY_ZIP_PATH, "r") as z:
        z.extractall(BINARY_DIR)

    return "/tmp/promtail-linux-amd64"
    

def _obtain_promtail(self, event) -> None:
    if self._is_promtail_binary_in_workload():
        return
    
    bin_path = self._download_promtail(event)
    
    with open(bin_path, "rb") as f:
        self._container.push(WORKLOAD_BINARY_PATH, f, permissions=0o755)
```

You could even be a little racier, spend a bit of memory and reduce the amount of file IO you're doing, though this does assume at least ~75MB of free memory -- note however that right now Pebble uploads straight from memory either way, so if this approach isn't suitable, we may have wider problems:

```python
def _download_promtail(self, event) -> None:
  with urlopen(PROMTAIL_BINARY_ZIP_URL) as r:
      file_bytes = r.read()
      
      if sha256(file_bytes).hexdigest() != BINARY_SHA256SUM:
          logger.error(
            "promtail binary mismatch, expected:'{}' but got '{}'", 
            BINARY_SHA256SUM, result
          )
          raise PromtailDigestError("Digest mismatch for promtail binary")
      
      ZipFile(io.BytesIO(file_bytes)).extractall(BINARY_DIR)

  return "/tmp/promtail-linux-amd64"
```

Reply:
My first approach was like this one, having everything in a single method: `_download_promtail()`, but I changed to the actual approach because the method had more than one responsibility (download the file, check the sha256sum, and unzip it)
Anyway I have no strong opinions about both approaches.

Reply:
In the current approach there is just an excessive amount of IO from all the opening/closing/reading/writing files. Interested in @simskij opinion here.

Reply:
I get your point about the separation of concerns, @Abuelodelanada, but in this case I would suggest inlining your helper functions to make it a smoother read. If you also adopt @jnsgruk's suggestion with not writing the tarball to disk but rather store it in a file object,untar it in-memory and write the files you need to disk, the code will end up quite compact regardless.